pkg/coscheduling/coscheduling.go

```diff
@@ -0,0 +1,260 @@
+	podGroupInfos := make(map[string]PodGroupInfo)
+	return &Coscheduling{FrameworkHandle: handle,
+		PodLister:     podLister,
+		PodGroupInfos: podGroupInfos,
+	}, nil
+}
+
+// Less are used to sort pods in the scheduling queue.
+// 1.compare the priority of pod
+// 2.compare the timestamp of the initialization time of PodGroup
+// 3.compare the UID
+func (cs *Coscheduling) Less(podInfo1 *framework.PodInfo, podInfo2 *framework.PodInfo) bool {
+	pod1 := podInfo1.Pod
+	pod2 := podInfo2.Pod
+	priority1 := pod.GetPodPriority(pod1)
+	priority2 := pod.GetPodPriority(pod2)
+	if priority1 != priority2 {
+		return priority1 > priority2
+	}
+
+	var pgInfo1 PodGroupInfo
+	var pgInfo2 PodGroupInfo
+	var ok bool
+	podGroupName1, min1 := cs.checkPod(pod1)
+	podGroupName2, min2 := cs.checkPod(pod2)
+	if podGroupName1 != "" {
+		pgInfo1, ok = cs.PodGroupInfos[podGroupName1]
+		if !ok {
+			pgInfo1 = PodGroupInfo{
+				InitialTimestamp: pod1.CreationTimestamp.Time,
+				Id:               cs.GetId(),
+				MinAvailable:     min1,
+			}
+			cs.PodGroupInfos[podGroupName1] = pgInfo1
+		}
+	}
+	if podGroupName2 != "" {
+		pgInfo2, ok = cs.PodGroupInfos[podGroupName2]
+		if !ok {
+			pgInfo2 = PodGroupInfo{
+				InitialTimestamp: pod2.CreationTimestamp.Time,
+				Id:               cs.GetId(),
+				MinAvailable:     min2,
+			}
+			cs.PodGroupInfos[podGroupName2] = pgInfo2
+		}
+	}
+
+	if podGroupName1 != "" && podGroupName2 != "" {
+		if pgInfo1.InitialTimestamp.Equal(pgInfo2.InitialTimestamp) {
+			return pgInfo1.Id < pgInfo2.Id
+		} else {
+			return pgInfo1.InitialTimestamp.Before(pgInfo2.InitialTimestamp)
+		}
+	} else if podGroupName1 != "" {
+		return pgInfo1.InitialTimestamp.Before(podInfo2.InitialAttemptTimestamp)
+	} else if podGroupName2 != "" {
+		return podInfo1.InitialAttemptTimestamp.Before(pgInfo2.InitialTimestamp)
+	}
+
+	return podInfo1.InitialAttemptTimestamp.Before(podInfo2.InitialAttemptTimestamp)
+}
+
+func (cs *Coscheduling) PreFilter(ctx context.Context, state *framework.CycleState, p *v1.Pod) *framework.Status {
+	podGroupName, minAvailable := cs.checkPod(p)
+	if podGroupName == "" || minAvailable == 0 {
+		return framework.NewStatus(framework.Success, "")
+	}
+
+	total := cs.calculateTotalPods(podGroupName)
+	if total < minAvailable {
+		klog.V(3).Infof("The count of podGroup %v is not up to minAvailable(%d) in PreFilter: %d",
+			podGroupName, minAvailable, total)
+		return framework.NewStatus(framework.Unschedulable, "less than minAvailable")
+	}
+
+	return framework.NewStatus(framework.Success, "")
+}
+
+func (cs *Coscheduling) PreFilterExtensions() framework.PreFilterExtensions {
+	return nil
+}
+
+// Permit is the functions invoked by the framework at "permit" extension point.
+func (cs *Coscheduling) Permit(ctx context.Context, state *framework.CycleState, p *v1.Pod, nodeName string) (*framework.Status, time.Duration) {
+	podGroupName, minAvailable := cs.checkPod(p)
+	if podGroupName == "" || minAvailable == 0 {
```

Comment:
```suggestion
	if podGroupName == "" || minAvailable <= 1 {
```

could we shortcut `minAvailable==1` here??